ble_app.h

```diff
@@ -0,0 +1,427 @@
+ */
+
+#ifndef BLE_APP_H_
+#define BLE_APP_H_
+
+#include "pretty_printer.h"
+#include "ble/BLE.h"
+#include "ChainableGapEventHandler.h"
+#include "events/mbed_events.h"
+#include "platform/Callback.h"
+#include "platform/NonCopyable.h"
+
+static const uint16_t MAX_ADVERTISING_PAYLOAD_SIZE = 50;
+
+/**
+ * Handle initialization and shutdown of the BLE Instance.
+ * It will also run the  event queue and call your post init callback when everything is up and running.
+ */
+class BLEApp : private mbed::NonCopyable<BLEApp>, public ble::Gap::EventHandler
+{
+public:
+    /**
+     * Construct a BLEApp from an event queue and a ble interface.
+     * Call start() to initiate ble processing.
+     */
+    BLEApp() : _ble(BLE::Instance())
+    {
+    }
+
+    ~BLEApp()
+    {
+        stop();
+    }
+
+    /**
+     * Initialize the ble interface, this will call the callback at the end of init.
+     *
+     * @param post_init_cb Callback that will be called on init complete.
+     */
+    void start(mbed::Callback<void(BLE&, events::EventQueue&)> post_init_cb)
+    {
+        printf("Ble App started\r\n");
+
+        _post_init_cb = post_init_cb;
+
+        if (_ble.hasInitialized()) {
+            printf("Error: the ble instance has already been initialized.\r\n");
+            return;
+        }
+
+        /* handle gap events */
+        _gap_handler.addEventHandler(this);
+        _ble.gap().setEventHandler(&_gap_handler);
+
+        /* This will inform us off all events so we can schedule their handling
+         * using our event queue */
+        _ble.onEventsToProcess(
+            makeFunctionPointer(this, &BLEApp::schedule_ble_events)
+        );
+
+        ble_error_t error = _ble.init(
+            this, &BLEApp::on_init_complete
+        );
+
+        if (error) {
+            print_error(error, "Error returned by BLE::init.\r\n");
+            return;
+        }
+
+        /* Process the event queue. */
+        _event_queue.dispatch_forever();
+        /* run left over events */
+        _event_queue.dispatch_once();
+
+        return;
+    }
+
+    /**
+     * Close advertising and/or existing connections and stop the App.
+     */
+    void stop()
+    {
+        /* let any left over events run */
+        _event_queue.call([this]() {
+            if (_ble.hasInitialized()) {
+                _ble.shutdown();
+                printf("Ble App stopped.\r\n");
+            }
+            _event_queue.break_dispatch();
+
+            _connected = false;
+            _is_connecting = false;
+            _is_scanning = false;
+            _gap_handler = ChainableGapEventHandler();
+        });
+    }
+
+    /**
+     * Subscribe with your own gap handler.
+     *
+     * @param[in] gap_handler Handler implementing selected ble::Gap::EventHandler methods.
+     *
+     * @returns True on success.
+     */
+    bool add_gap_event_handler(ble::Gap::EventHandler *gap_handler)
+    {
+        return _gap_handler.addEventHandler(gap_handler);
```

Comment:
Chainable event handler needs a safeAddHandler function that first checks for duplicates.
There is a function to check if handler is present but it's currently broken.